source/agora/utils/TxBuilder.d

```diff
@@ -332,7 +388,54 @@ public struct TxBuilder
 
         // Sign all inputs using unlocker
         foreach (idx, ref in_; this.inputs)
-            this.data.inputs[idx].unlock = unlocker(this.data, in_);
+            this.data.inputs[idx].unlock = faucet_unlocker(this.data, in_);
+
+        // Reset ready for next time
+        this.inputs = null;
+        this.leftover = Output.init;
+        // Reset transaction if it is returned successfully
+        scope (success) this.data = Transaction.init;
+        return this.data;
+    }
+
+    public Transaction sign (
+        in OutputType outputs_type = OutputType.Payment, uint unlock_age = 0) @safe nothrow
+    {
+        assert(this.inputs.length, "Cannot sign input-less transaction");
+        assert(this.data.outputs.length || this.leftover.value > Amount(0),
+               "Output-less transactions are not valid");
+
+        // First we sort the OutputRefs as later we set the unlock by index
+        this.inputs.sort;
+        // Add the inputs with just their unlocks to be added
+        // (unlock is not part of Transaction hash but we need transaction hash to create the unlock)
+        this.inputs.each!(o => this.data.inputs ~= Input(o.hash, Unlock.init, unlock_age));
+
+        assert(this.data.inputs.isStrictlyMonotonic);
+
+        foreach (ref o; this.data.outputs)
+            o.type = outputs_type;
+
+        auto total_fees = this.minFees();
+        assert(this.leftover.value >= total_fees);
+        // Add the refund output if more than `MinRefundAmount` is leftover
+        if (this.leftover.value > total_fees)
```

Comment:
This is missing something no ?

Reply:
If it is equal then it should not add anything. Is that what you mean?

Reply:
Unless I'm missing something, we're not comparing to `this.MinRefundAmount` so we'll create a refund output regardless of its value ?

Reply:
Or maybe you mean `if (this.leftover.value > total_fees + this.MinRefundAmount)` which is a bit clearer.

Reply:
But then it is not `nothrow`.

Reply:
Down three lines we do.

Reply:
I refactored to make it clearer anyway.